src/core/algorithms/sfd/sfd_algorithm.cpp

```diff
@@ -0,0 +1,302 @@
+    using config::Option;
+    auto check_param = [](long double param) {
+        if (param < 0 || param > 1) throw config::ConfigurationError("Parameter out of range");
+    };
+    auto check_max_false_positive = [](long double param) {
+        if (param < 0 || param > 0.39)
+            throw config::ConfigurationError(
+                    "Maximum probability of a false-positive result is out of range");
+    };
+    auto check_positive = [](size_t param) {
+        if (param <= 0) throw config::ConfigurationError("Parameter out of range");
+    };
+    auto check_delta = [this](long double param) {
+        if (param < 0 || param > 1) throw config::ConfigurationError("delta out of range");
+        if (param < minimum_cardinality_) {
+            throw config::ConfigurationError("delta must be less than minimum_cardinality_");
+        }
+    };
+    RegisterOption(config::kTableOpt(&input_table_));
+    RegisterOption(Option{&only_sfd_, kOnlySFD, kDOnlySFD});
+    RegisterOption(Option{&minimum_cardinality_, kMinCard, kDMinCard}.SetValueCheck(check_param));
+    RegisterOption(
+            Option{&max_diff_vals_proportion_, kMaxDiffValsProportion, kDMaxDiffValsProportion}
+                    .SetValueCheck(check_param));
+    RegisterOption(
+            Option{&min_sfd_strength_measure_, kMinSFDStrengthMeasure, kDMinSFDStrengthMeasure}
+                    .SetValueCheck(check_param));
+    RegisterOption(
+            Option{&min_skew_threshold_, kMinSkewThreshold, kDMinSkewThreshold}.SetValueCheck(
+                    check_param));
+    RegisterOption(Option{&min_structural_zeroes_amount_, kMinStructuralZeroesAmount,
+                          kDMinStructuralZeroesAmount}
+                           .SetValueCheck(check_param));
+    RegisterOption(Option{&max_false_positive_probability_, kMaxFalsePositiveProbability,
+                          kDMaxFalsePositiveProbability}
+                           .SetValueCheck(check_max_false_positive));
+
+    RegisterOption(Option{&delta_, kDelta, kDDelta}.SetValueCheck(check_delta));
+    RegisterOption(
+            Option{&max_amount_of_categories_, kMaxAmountOfCategories, kDMaxAmountOfCategories}
+                    .SetValueCheck(check_positive));
+}
+
+void SFDAlgorithm::MakeExecuteOptsAvailable() {
+    using namespace config::names;
+    MakeOptionsAvailable({kOnlySFD, kMinCard, kMaxDiffValsProportion, kMinSFDStrengthMeasure,
+                          kMinSkewThreshold, kMinStructuralZeroesAmount,
+                          kMaxFalsePositiveProbability, kDelta, kMaxAmountOfCategories});
+}
+
+void SFDAlgorithm::ResetState() {
+    is_skewed_.clear();
+    domains_.clear();
+    soft_keys_.clear();
+    trivial_columns_.clear();
+    correlations_.clear();
+    unrelated_.clear();
+    sfd_.clear();
+}
+
+void SFDAlgorithm::LoadDataInternal() {
+    typed_relation_ = model::ColumnLayoutTypedRelationData::CreateFrom(*input_table_, false);
```

Comment:
> Can you explain why we use `false` here?
Actually original paper doesn’t give any guidelines whether two distinct occurrences of a NULL should be considered the same value. In case sofd fd between columns is not present the next step will be test for correlation. Considering NULLS as distinct values will avoid skew of distribution caused by NULLS  and they will be filtered out before the correlation test if distribution is skewed, so as i was writing this piece of code it seemed reasonable to me to view NULLS as distinct. But now i am not sure about this because such approach affects detection of soft fds as it uses amount of distinct values in a column. Maybe it would be better to make new param for the algorithm `is_null_eq_null` to let the user decide whether NULLS are equal or not.